File: packages/react-native-messaging-core/ios/Sources/Models/Response.swift
import Foundation
import NablaCore

extension RefreshingState {
    var dictionaryRepresentation: [String: Any] {
        switch self {

        case .refreshing:
            return ["type": "Refreshing"]
        case .refreshed:
            return ["type": "Refreshed"]
        case let .failed(error):
            return [
                "type": "ErrorWhileRefreshing",
                "error": error.dictionaryRepresentation
            ]
        }
    }
}

extension Response {
    func dictionaryRepresentation(
        _ dataDictionaryRepresentation: (Data) -> [String: Any]
    ) -> [String: Any] {
        [
            "isDataFresh": isDataFresh,
            "refreshingState": refreshingState.dictionaryRepresentation,
            "data": dataDictionaryRepresentation(data)

        ]
    }
}

File: packages/react-native-messaging-core/src/types/Response.ts
export type RefreshingState<Error> =
  | { type: 'Refreshing' }
  | { type: 'Refreshed' }
  | { type: 'ErrorWhileRefreshing'; error: Error };

export type Response<Data, Error> = {
  isDataFresh: boolean;
  refreshingState: RefreshingState<Error>;
  data: Data;
};
